### PR TITLE
[flutter_tools] cache-bust in service worker

### LINKED ISSUE #56532: [web] should use the version settled in pubspec.yaml
As I am building more and more flutter web apps, I ran into the same problem.

When I deploy my app with Firebase Hosting, the customer doesn't see the newest version. He has to delete his cache first. 

> The solution to a problem lies in the problem itself.

The web browser doesn't detect, that there is a new version online. 

Quick Google search and you find the following:

[Caching in Flutter for Web](https://medium.com/flutter-community/caching-in-flutter-for-web-42b3ae0e348f)
Thank you [Constantin Stan](https://medium.com/@constanting)

We only have to add a version in our **index.html**

```
<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <title>to_cache_or_not_to_cache</title>
  </head>
  <body>
    <script src="main.dart.js?version=1" type="application/javascript"></script>
  </body>
</html>
```

I would like Flutter to be able to do it by itself with the version settled in the `pubspec.yaml`

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Prevent HTTP cache from breaking invariants of service worker cache by forcing a new version through no-cache param. Fixes issue where a fetch request for `/#/` or other routes would not redirect to index.html

Also adds the functionality required for https://github.com/flutter/packages/pull/150

Fixes https://github.com/flutter/flutter/issues/56532

### CODE REVIEW

**Comment by jonahwilliams** [comment]
I was thinking about an integration test for this, We spin up a dart server to serve the app and then verify that http requests from it

**Comment by jonahwilliams** [comment]
@yjbanov do you have any advice on debugging why chrome won't start up? I can get these to pass locally with `headless:false`,but I always get "Exited with unexpected code 0" on headless:true

File: dev/bots/test.dart [hunk 1]
```diff
@@ -851,6 +853,75 @@ Future<void> _runWebReleaseTest(String target, {
...
+  final List<String> expectedPaths = <String>[
+    // Initial page load
+    '',
+    'main.dart.js',
+    'assets/FontManifest.json',
+    'flutter_service_worker.js',
+    'manifest.json',
+    'favicon.ico',
+    // Service worker install.
+    'main.dart.js?id=18f7de7f7ebd8ff11a8e835cd99d45bc',
```
Comment (jonahwilliams) [inline dev/bots/test.dart:905]: Note: I need a better solution here since this will change every time dart rolls

**Review by yjbanov — APPROVED** [review]

**Review by yjbanov — APPROVED** [review]

**Comment by sgehrman** [comment]
I don't see this working.  Maybe I'm misunderstanding.  I change my flutter web project, rebuild it and deploy it, but refreshing the browser doesn't get the new build.  I have to clear the application data.  How do I get this working automatically?  Do I have to add a version in index.html for the js file?

**Comment by jeffaknine** [comment]
I'm having the same issue, can we get a follow up on this? @jonahwilliams

**Pull request merged**